botensure.go

```diff
@@ -0,0 +1,172 @@
+package pluginapi
+
+import (
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/blang/semver"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+	"github.com/mattermost/mattermost-server/v5/utils"
+	"github.com/pkg/errors"
+)
+
+type ensureBotOptions struct {
+	ProfileImagePath string
+	IconImagePath    string
+}
+
+type EnsureBotOption func(*ensureBotOptions)
+
+func (b *BotService) ensureServerVersion(required string) error {
+	serverVersion := b.api.GetServerVersion()
+	currentVersion := semver.MustParse(serverVersion)
+	requiredVersion := semver.MustParse(required)
+
+	if currentVersion.LT(requiredVersion) {
+		return errors.Errorf("incompatible server version for plugin, minimum required version: %s, current version: %s", required, serverVersion)
+	}
+	return nil
+}
+
+// EnsureBot either returns an existing bot user matching the given bot, or creates a bot user from the given bot.
+// A profile image or icon image may be optionally passed in to be set for the existing or newly created bot.
+// Returns the id of the resulting bot.
+//
+// Minimum server version: 5.10
+func (b *BotService) EnsureBot(bot *model.Bot, options ...EnsureBotOption) (retBotID string, retErr error) {
+	err := b.ensureServerVersion("5.10.0")
+	if err != nil {
+		return "", errors.Wrap(err, "failed to ensure bot")
+	}
+
+	// Default options
+	o := &ensureBotOptions{
+		ProfileImagePath: "",
+		IconImagePath:    "",
+	}
+
+	for _, setter := range options {
+		setter(o)
+	}
+
+	botID, err := b.ensureBot(bot)
+	if err != nil {
+		return "", err
+	}
+
+	err = b.setBotImages(botID, o.ProfileImagePath, o.IconImagePath)
+	if err != nil {
+		return "", err
+	}
+	return botID, nil
+}
+
+func (b *BotService) ensureBot(bot *model.Bot) (retBotID string, retErr error) {
+	// Must provide a bot with a username
+	if bot == nil || len(bot.Username) < 1 {
+		return "", errors.New("passed a bad bot, nil or no username")
+	}
+
+	// If we fail for any reason, this could be a race between creation of bot and
+	// retrieval from another EnsureBot. Just try the basic retrieve existing again.
+	defer func() {
+		if retBotID == "" || retErr != nil {
+			var err error
+			var botIDBytes []byte
+
+			err = utils.ProgressiveRetry(func() error {
```

Comment:
I know it's just copied, but depending on `utils` from mattermost-server is annoying. Perhaps something to port over eventually as well.

Reply:
Made ticket: https://mattermost.atlassian.net/browse/MM-23501